src/webots/nodes/WbBallJoint.cpp

```diff
@@ -74,64 +323,173 @@ bool WbBallJoint::setJoint() {
+  switch (index) {
+    case 1:
+      return mInitialPosition;
+    case 2:
+      return mInitialPosition2;
+    case 3:
+      return mInitialPosition3;
+    default:
+      return NAN;
+  }
+}
+
+void WbBallJoint::setPosition(double position, int index) {
+  if (index == 3) {
+    mPosition3 = position;
+    mOdePositionOffset3 = position;
+    WbJointParameters *const p3 = parameters3();
+    if (p3)
+      p3->setPosition(mPosition3);
+
+    WbMotor *const m3 = motor3();
+    if (m3)
+      m3->setTargetPosition(position);
+    return;
+  }
+  WbHinge2Joint::setPosition(position, index);
+}
+
+bool WbBallJoint::resetJointPositions() {
+  mOdePositionOffset3 = 0.0;
+  return WbHinge2Joint::resetJointPositions();
+}
+
+void WbBallJoint::updateOdePositionOffset() {
+  WbHinge2Joint::updateOdePositionOffset();
+  mOdePositionOffset3 = position(3);
 }
 
 void WbBallJoint::preFinalize() {
-  WbBasicJoint::preFinalize();
+  WbHinge2Joint::preFinalize();
 
   WbBallJointParameters *const p = ballJointParameters();
   if (p && !p->isPreFinalizedCalled())
     p->preFinalize();
+
+  for (int i = 0; i < devices3Number(); ++i) {
+    if (device3(i) && !device3(i)->isPreFinalizedCalled())
+      device3(i)->preFinalize();
+  }
+  // the following code could be simplified
+  WbBaseNode *const p3 = dynamic_cast<WbBaseNode *>(mParameters3->value());
+  if (p3 && !p3->isPreFinalizedCalled())
+    p3->preFinalize();
+
+  updateParameters3();
+  checkMotorLimit();
+
+  mInitialPosition3 = mPosition3;
 }
 
 void WbBallJoint::postFinalize() {
-  WbBasicJoint::postFinalize();
+  WbHinge2Joint::postFinalize();
 
   WbBallJointParameters *const p = ballJointParameters();
 
   if (p && !p->isPostFinalizedCalled())
     p->postFinalize();
-}
-
-void WbBallJoint::updateAnchor() {
-  if (mJoint)
-    applyToOdeAnchor();
-
-  if (WbWrenRenderingContext::instance()->isOptionalRenderingEnabled(WbWrenRenderingContext::VF_JOINT_AXES))
-    updateJointAxisRepresentation();
-}
-
-void WbBallJoint::applyToOdeAnchor() {
-  const WbMatrix4 &m4 = upperTransform()->matrix();
-  const WbVector3 &t = m4 * anchor();
-  dJointSetBallAnchor(mJoint, t.x(), t.y(), t.z());
+  for (int i = 0; i < devices3Number(); ++i) {
+    if (device3(i) && !device3(i)->isPostFinalizedCalled())
+      device3(i)->postFinalize();
+  }
+  // the following code could be simplified
+  WbBaseNode *const p3 = dynamic_cast<WbBaseNode *>(mParameters3->value());
+  if (p3 && !p3->isPostFinalizedCalled())
+    p3->postFinalize();
+
+  connect(mDevice3, &WbMFNode::itemInserted, this, &WbBallJoint::addDevice3);
+  connect(mParameters3, &WbSFNode::changed, this, &WbBallJoint::updateParameters);
+  if (p)
+    connect(p, &WbBallJointParameters::anchorChanged, this, &WbBallJoint::updateAnchor, Qt::UniqueConnection);
+  if (brake3())
+    connect(brake3(), &WbBrake::brakingChanged, this, &WbBallJoint::updateSpringAndDampingConstants, Qt::UniqueConnection);
+  if (motor2()) {
```

Comment:
Shouldn't it be motor3 here?

Reply:
No, only the second motor range is limited, not the third motor.